Move notification listening to MainActivity

madim opened the pull request:
Move notification listening to MainActivity
Avoid false positive notifications for observable queries
Remove unused files

madim committed:
Move notification listening to MainActivity

Avoid false positive notifications for observable queries
Remove unused files

madim commented on app/src/main/java/org/covidwatch/android/CurrentUserExposureNotifier.kt:
```diff
@@ -1,74 +0,0 @@
-
-        createNotificationChannel()
-
-        viewModel.contactEvents.observeForever(Observer {
```
Deleted this implementation of UserExposureNotifier because here we might have a memory leak because we are not removing observer.

nkumarcc commented on app/src/main/java/org/covidwatch/android/di/AppModule.kt:
```diff
@@ -28,6 +29,13 @@ val appModule = module {
 
+    viewModel {
+        MainViewModel(
+            testRepository = get(),
```
Can we rename `testRepository` at some point?

madim replied:
yeah for sure! you have something in your mind?

nkumarcc replied:
Honestly I'd prefer `Tested`, `IsTested`, or `IsTestedPositiver`, just because `Test` can be ambiguous. A little weird still but more specific

madim replied:
Renamed to `Tested`

nkumarcc reviewed (approved):
My thoughts: looks fine, is there a reason we need `observeForever` + the remove there instead of just `observe`? Also, what's the purpose of the `CovidWatchApplication` file at this point, can we deprecate it?

nkumarcc commented:
"Avoid false positive notifications for observable queries" which notifs were these?

madim committed:
Removed Worker that downloaded contact_events from firebase on every launch of the app

Made sure that user sees possible exposure notification even if app is in background

madim commented:
> My thoughts: looks fine, is there a reason we need `observeForever` + the remove there instead of just `observe`? Also, what's the purpose of the `CovidWatchApplication` file at this point, can we deprecate it?

We can only use `observeForever` inside ViewModel, because there is nothing we can attach it to. We just need to make sure that we remove it in `onCleared()`.

There are still some logic `CovidWatchApplication` so we can't remove it right now.

madim committed:
Rename TestRepository to TestedRepository

nkumarcc commented on app/src/main/java/org/covidwatch/android/data/contactevent/ContactEventsDownloadWorker.kt:
```diff
@@ -8,11 +8,17 @@ import org.covidwatch.android.R
+    private val context: Context,
+    workerParams: WorkerParameters
+) : Worker(context, workerParams),
+    KoinComponent {
```
Do you need to add this to add to the `DI` chain so it can be injected? Or just allow it to use injectable things

madim replied:
Yep, from docs:
>KoinComponent interface marker to bring Koin extensions features

However koin has already marked most of the android components (Application, Activity, Fragment) so we don't to mark them manually

nkumarcc commented on app/src/main/java/org/covidwatch/android/presentation/MainActivity.kt:
```diff
@@ -45,13 +46,7 @@ class MainActivity : AppCompatActivity() {
     }
 
     private fun notifyUserOfPossibleExposure() {
```
Is this function not a duplicate of the functionality in `NotifyAboutPossibleExposureUseCase.kt`?

madim replied:
Yes it is, but we might need it to show notification when firebase triggers a change in https://github.com/covid19risk/covidwatch-android/pull/90.

nkumarcc commented:
Bump on "Avoid false positive notifications for observable queries" which notifs were these?

Also, looks like you just dropped the one time loader. Guessing this fixed it?

nkumarcc commented:
Can we also change the 3 hr window to 1 hr?

madim commented:
> Bump on "Avoid false positive notifications for observable queries" which notifs were these?
> 
> Also, looks like you just dropped the one time loader. Guessing this fixed it?

No(
If you remove `getDistinct()` [here](https://github.com/covid19risk/covidwatch-android/pull/93/files#diff-6568a7d63439f105243e3c7686bfb56bR41), set breakpoint at this [line](https://github.com/covid19risk/covidwatch-android/pull/93/files#diff-6568a7d63439f105243e3c7686bfb56bR44), and debug the app, you'll see that it triggers 5(!) times

You can read here why that happens:
https://medium.com/androiddevelopers/7-pro-tips-for-room-fbadea4bfbd1
>7. Avoid false positive notifications for observable queries

madim committed:
Remove MainViewModel. Change 3 hour PeriodicWorkRequestBuilder to 1 hour

nkumarcc commented:
Ahhhh yes, I've seen the distinct bug, thought it might be some weird BLE or Firebase bug. Good catch
